test/integration/test_integration.py

```diff
@@ -11,54 +14,69 @@
 Result = namedtuple('Result', ['outcome', 'artifact', 'item'])
 
 
-def parse_results(log):
-    '''
-    From the given stdout log, parse the results
-    '''
-    start = 'results:'
-    results = []
-    record = False
+class MockEnv:
+    '''Use this to work with mock. Mutliple instances are not safe.'''
+
+    def __init__(self):
+        self.mock = ['mock', '-r', './mock.cfg']
+        self._run(['--init'], check=True)
+
+    def _run(self, what, **kwargs):
+        return subprocess.run(self.mock + what, **kwargs)
+
+    def copy_in(self, files):
+        self._run(['--copyin'] + files + ['/'], check=True)
+
+    def copy_out(self, directory, *, clean_target=False):
+        if clean_target:
+            with contextlib.suppress(FileNotFoundError):
+                shutil.rmtree(directory)
+        self._run(['--copyout', directory, directory], check=True)
+
+    def shell(self, command):
+        self._run(['--shell', command])
+
+    def orphanskill(self):
+        self._run(['--orphanskill'])
 
-    for line in log.splitlines():
-        if line.strip() == start:
-            record = True
-        if record:
-            results.append(line)
-            if not line:
-                break
 
-    if not results:
-        raise RuntimeError('Could not parse output')
-    return yaml.load('\n'.join(results))['results']
+@pytest.fixture(scope="session")
+def mock():
+    '''Setup a mock we can run Ansible tasks in under root'''
+    mockenv = MockEnv()
+    files = ['taskotron_python_versions'] + glob.glob('*.py') + ['tests.yml']
+    mockenv.copy_in(files)
+    yield mockenv
+    mockenv.orphanskill()
 
 
-def run_task(nevr, *, reterr=False):
+def parse_results(path):
     '''
-    Run the task on a Koji build.
+    From the given result file, parse the results
+    '''
+    with open(path) as f:
+        return yaml.load(f)['results']
+
+
+def run_task(nevr, *, mock):
+    '''
+    Run the task on a Koji build in given mock.
     Returns a dict with Results (outcome, artifact, item)
-    If reterr is true, returns a tuple with the above and captured stderr
-    If reterr is false, prints the stderr
+    Actually returns a tuple with the above and captured log
     '''
-    proc = subprocess.Popen(
-        ['runtask', '-i', nevr, '-t', 'koji_build', 'runtask.yml'],
-        stderr=subprocess.PIPE,
-        universal_newlines=True,
-    )
-    _, err = proc.communicate()
-    if proc.returncode != 0:
-        print(err, file=sys.stderr)  # always print stderr in this case
-        raise RuntimeError('runtask exited with {}'.format(proc.returncode))
-    results = parse_results(err)
+    mock.shell('ansible-playbook tests.yml -e taskotron_item={}'.format(nevr))
```

Comment:
This ends up with positive exit code when the check FAIL. So I didn't check the result. However there is failure (of the check) and failure (of the entire circus) and we should check the exit status and try to guess (at least if it's negative, it was killed and we should raise KeybordInterupt or something like that).

Reply:
Added, it's either 0 or 2.

Reply:
To clarify: the SI spec says the playbook exit code reflects the result of the check - 0 means PASS, non-zero means FAIL. It doesn't know the concept of generic tests. To stay close to the SI spec, we opted (at least for now) to exiting with 0 if all tests pass and non-zero if at least one check fails in our generic tasks. Unfortunately that prevents us from distinguishing a failed test from a crashed test. So, as long there is `{{artifacts}}/taskotron/results.yml` present, we consider the test to have finished fine. This is something we should discuss with SI spec authors in the future.